test/test_grid_file.py

```diff
@@ -447,27 +462,85 @@ def test_readline(self):
         self.assertEqual(b"e", g.readline(1))
         self.assertEqual(b"llo world,\n", g.readline())
 
+    def test_readlines(self):
+        f = GridIn(self.db.fs, chunkSize=5)
+        f.write((b"""Hello world,
+How are you?
+Hope all is well.
+Bye"""))
+        f.close()
+
+        # Try read(), then readlines().
+        g = GridOut(self.db.fs, f._id)
+        self.assertEqual(b"He", g.read(2))
+        self.assertEqual([b"llo world,\n", b"How are you?\n"], g.readlines(11))
+        self.assertEqual([b"Hope all is well.\n", b"Bye"], g.readlines())
+        self.assertEqual([], g.readlines())
+
+        # Try readline(), then readlines().
+        g = GridOut(self.db.fs, f._id)
+        self.assertEqual(b"Hello world,\n", g.readline())
+        self.assertEqual([b"How are you?\n", b"Hope all is well.\n"], g.readlines(13))
+        self.assertEqual(b"Bye", g.readline())
+        self.assertEqual([], g.readlines())
+
+        # Only readlines().
+        g = GridOut(self.db.fs, f._id)
+        self.assertEqual(
+            [b"Hello world,\n", b"How are you?\n", b"Hope all is well.\n", b"Bye"],
+            g.readlines())
+
+        g = GridOut(self.db.fs, f._id)
+        self.assertEqual(
+            [b"Hello world,\n", b"How are you?\n", b"Hope all is well.\n", b"Bye"],
+            g.readlines(0))
+
+        g = GridOut(self.db.fs, f._id)
+        self.assertEqual([b"Hello world,\n"], g.readlines(1))
+        self.assertEqual([b"How are you?\n"], g.readlines(12))
+        self.assertEqual([b"Hope all is well.\n", b"Bye"], g.readlines(18))
+
+        # Try readlines() first, then read().
+        g = GridOut(self.db.fs, f._id)
+        self.assertEqual([b"Hello world,\n"], g.readlines(1))
+        self.assertEqual(b"H", g.read(1))
+        self.assertEqual([b"ow are you?\n", b"Hope all is well.\n"], g.readlines(29))
+        self.assertEqual([b"Bye"], g.readlines(1))
+
+        # Try readlines() first, then readline().
+        g = GridOut(self.db.fs, f._id)
+        self.assertEqual([b"Hello world,\n"], g.readlines(1))
+        self.assertEqual(b"How are you?\n", g.readline())
+        self.assertEqual([b"Hope all is well.\n"], g.readlines(17))
+        self.assertEqual(b"Bye", g.readline())
+
     def test_iterator(self):
         f = GridIn(self.db.fs)
         f.close()
         g = GridOut(self.db.fs, f._id)
         self.assertEqual([], list(g))
 
         f = GridIn(self.db.fs)
-        f.write(b"hello world")
+        f.write(b"hello world\nhere are\nsome lines.")
```

Comment:
This test surprised me when I made changes to it. It previously tested:

```python
g = GridOut(self.db.fs, f._id)
self.assertEqual([b"hello world"], list(g))
self.assertEqual(b"hello", g.read(5))
self.assertEqual([b"hello world"], list(g))
self.assertEqual(b" worl", g.read(5))
```

so iterating with GridOutIterator previously did not actually move the "file pointer" ahead and doing
`next(iter(g))` always gives the same result (first chunk of the file).

The IOBase implementation moves the file pointer ahead when iterating, so expected behaviour is e.g.

```python
>>> with open("test.txt", "w") as f_out:
...     f_out.write("here are\nsome lines\nin a file")
>>> with open("test.txt", "r") as f_in:
...     print(next(iter(f_in)))
...     print(f_in.read(5))
...     print(list(f_in))
... 
here are

some 
['lines\n', 'in a file']
```

and that's also the behaviour that GridOut iter has with this PR. Is there something I'm missing as to why the iterator previously always started from the top and didn't move the "file pointer" ahead?

Reply:
Great catch. I don't understand why we had that behavior. The new behavior seems much more reasonable. I'm running the test suite to see if any other regression tests break. If the tests pass then I think we're good to go. We can make a minor note in the changelog that users can call `seek(0)` to reset the iterator.

Reply:
Will add a note to the changelog